05_classify_mutations_multimodal/nbconverted/plot_multimodal_results.py

```diff
@@ -0,0 +1,304 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# ## Plot results of multi-omics classification experiments
+# 
+# In these experiments, we compare elastic net logistic regression models using multiple data types to models using only a single data type. We're not doing anything particularly fancy here, just concatenating the feature sets (genes or CpG probes) from the individual data types to create a "multi-omics" model.
+# 
+# For now, we're just doing this for gene expression and the two methylation datasets.
+
+# In[1]:
+
+
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import matplotlib.patches as mpatches
+import seaborn as sns
+from scipy.stats import ttest_ind
+from adjustText import adjust_text
+
+import mpmp.config as cfg
+import mpmp.utilities.analysis_utilities as au
+
+
+# In[2]:
+
+
+# set results directory
+results_dir = Path(
+    cfg.results_dirs['multimodal'],
+    # 'pilot_results',
+    'pilot_results_all_feats',
+    'gene'
+).resolve()
+
+
+# ### Compare raw results (signal and shuffled)
+
+# In[3]:
+
+
+# load raw data
+results_df = au.load_stratified_prediction_results(results_dir, 'gene')
+
+# make sure that we have data for all data types and for two replicates (random seeds)
+print(results_df.shape)
+print(results_df.seed.unique())
+print(results_df.identifier.unique())
+print(results_df.training_data.unique())
+results_df.head()
+
+
+# In[4]:
+
+
+# each subplot will show results for one gene
+sns.set({'figure.figsize': (24, 12)})
+fig, axarr = plt.subplots(2, 4)
+results_df.sort_values(by=['identifier', 'signal', 'training_data'], inplace=True)
+max_aupr = results_df.aupr.max()
+
+data_order =['expression.me_27k',
+             'expression.me_450k',
+             'me_27k.me_450k',
+             'expression.me_27k.me_450k']
+
+# plot mean performance over all genes in Vogelstein dataset
+for ix, gene in enumerate(results_df.identifier.unique()):
+    
+    ax = axarr[ix // 4, ix % 4]
+    
+    plot_df = results_df[(results_df.identifier == gene) &
+                         (results_df.data_type == 'test')]
+
+    sns.boxplot(data=plot_df, x='signal', y='aupr', hue='training_data',
+                hue_order=data_order, ax=ax)
+    ax.set_title('Prediction for {} mutation'.format(gene))
+    ax.set_xlabel('')
+    ax.set_ylabel('AUPR')
+    ax.set_ylim(-0.1, max_aupr)
+    ax.legend_.remove()
+    
+handles = []
+for ix, data in enumerate(data_order):
+    handle = mpatches.Patch(color=sns.color_palette()[ix], label=data)
+    handles.append(handle)
+    
+fig.delaxes(axarr[1, 3])
+plt.legend(handles=handles, loc='lower right')
```

Comment:
just as a dumb sanity check, when you do the multi-modal models the performance is always increasing on the training set, right? I'm just surprised that the red boxplot is significantly below the blue for EGFR and PIK3CA. It would mean that the methylation data really has no information for those genes.

Reply:
This is a good question, thanks for pointing this out. Sorry in advance for the wall of text - I think this is important to think about.

Here's a plot of performance on the training set:

![image](https://user-images.githubusercontent.com/2345877/113923341-70f6da00-97b6-11eb-9bd4-7766d13943a1.png)

And the corresponding performance on the test set:

![image](https://user-images.githubusercontent.com/2345877/113923424-866c0400-97b6-11eb-98cf-716b86f05283.png)

So my answer to your question would be "sometimes but not necessarily". In some cases performance does increase on the training set but not the test set (e.g. PIK3CA, SETD2), but in other cases (EGFR, IDH1) performance is lower on both the train and test sets for the multi-modal models.

I'm not sure whether or not to be concerned about this, though. The way that I'm choosing hyperparameters for the elastic net model is by nested cross-validation (e.g. run 3-fold cross-validation on the training set with each set of hyperparameters), then I choose the best set of hyperparameters based on mean validation set AUPR.

That is to say that if I was doing something like standard ordinary least squares I would expect training set performance to always increase as more features are added, but since I choose the model that performs the best on the validation set, it seems plausible that the best model could be one that underfits the training set slightly but performs well on held-out data, as opposed to one that (over)fits the training set more closely and performs slightly less well on held-out data.

But I think it's important to ask whether we're really finding a good model (and, like you said, the added data really isn't contributing anything), or if we're just underfitting to the multi-omics data or not choosing the right optimizer or hyperparameter ranges for the larger data dimension. I can think of a few ways to get at this:

* Check if for the _same set_ of hyperparameters, our training set error is lower for multi-omics data than for single-omics data (I think this should be true)
* Try running the same experiments with slightly larger hyperparameter ranges, or a different optimizer

If you have any other ideas, I'd definitely be interested in hearing them (if you made it this far). I created #41 to look more closely at this, and I think it's important to spend some time looking at this before we submit the paper; we really don't want to conclude that multi-modal data doesn't help if it's really just a model fitting/optimization issue.

Reply:
> Check if for the same set of hyperparameters, our training set error is lower for multi-omics data than for single-omics data (I think this should be true)

i think this is a good approach, I also think this is true

> Try running the same experiments with slightly larger hyperparameter ranges, or a different optimizer

I think you can make a plot here to show loss vs hyperparamter value to see if you are finding the correct sweet-spot.  

I think it's an important aspect to make sure that you really identify or empirically show in some way that your model fit is optimal, I can just see it as a sticking point for a reviewer. I can't think of other ways to solidify it, but I'll let you know if I see something that could help.

Reply:
Great, thanks, this is super helpful. I'm going to merge this PR, but I'll look at this more closely in another PR in the near future.